### PR TITLE
Potential bug fixes on shell scripts

### PR DESCRIPTION
**Pull request by mmphego**
- Use $(..) instead of legacy `..`.
- Double quote to prevent globbing and word splitting.
- Redirecting `echo` output to `stderr`.
- Checking if `flake8`, `cpplint` and `isort` Python modules are installed.
- Using 'cd ... || exit' or 'cd ... || return' in case cd fails.
- Double quote to prevent globbing and word splitting.
- The use of > is for string comparisons. Using `-gt` instead.
- [Useless Use Of Cat](https://en.wikipedia.org/wiki/Cat_(Unix)#Useless_use_of_cat)
- Using echo -e "\n", instead of echo ""
- Using `find` instead of `ls` to better handle non-alphanumeric filenames

Signed-off-by: Mpho Mphego <mpho112@gmail.com>

### CODE REVIEW

File: ci/check_cache.sh [hunk 1]
```diff
@@ -4,24 +4,23 @@
...
+CACHE_DIR="${HOME}/.cache/"
+CCACHE_DIR="${HOME}/.ccache/"
 
-if [ $ci_changes -ne 0 ]
-then
+if [ "${ci_changes}" -ne 0 ]; then
     echo "Files have changed in ci/ deleting all caches"
-    rm -rf "$CACHE_DIR"
-    rm -rf "$CCACHE_DIR"
+    rm -rf "${CACHE_DIR}" || exit 1
```
Comment (jreback) [inline ci/check_cache.sh]: remove these

File: ci/circle/install_circle.sh [hunk 2]
```diff
@@ -27,55 +26,51 @@ conda info -a || exit 1
...
+    export LANG="${LOCALE_OVERRIDE}"
+    export LC_ALL="${LOCALE_OVERRIDE}"
 
-    python -c "import locale; locale.setlocale(locale.LC_ALL, \"$LOCALE_OVERRIDE\")" || exit 1;
+    python -c "import locale; locale.setlocale(locale.LC_ALL, \"${LOCALE_OVERRIDE}\")" || exit 1;
 
     echo "[Adding locale to the first line of pandas/__init__.py]"
-    rm -f pandas/__init__.pyc
-    sedc="3iimport locale\nlocale.setlocale(locale.LC_ALL, \"$LOCALE_OVERRIDE\")\n"
+    rm -f pandas/__init__.pyc || true;
```
Comment (jreback) [inline ci/circle/install_circle.sh]: you don't need the error conditions here

File: ci/code_checks.sh [hunk 3]
```diff
@@ -33,56 +33,63 @@ function invgrep {
...
 
 ### LINTING ###
-if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
+if [[ -z "${CHECK}" || "${CHECK}" == "lint" ]]; then
 
     # `setup.cfg` contains the list of error codes that are being ignored in flake8
 
-    echo "flake8 --version"
-    flake8 --version
+    # Check if flake8 is installed.
```
Comment (jreback) [inline ci/code_checks.sh]: you can do the check on flake8 w/o moving the code itself, eg. revert lines 62 and on.
if the command doesn't exist then need to print a message and immediately exit

File: ci/code_checks.sh [hunk 4]
```diff
@@ -33,56 +33,63 @@ function invgrep {
...
 
-    MSG='Linting code-blocks in .rst documentation' ; echo $MSG
-    flake8-rst doc/source --filename=*.rst --format="$FLAKE8_FORMAT"
-    RET=$(($RET + $?)) ; echo $MSG "DONE"
+        MSG='Linting code-blocks in .rst documentation' ; echo "${MSG}"
+        flake8-rst doc/source --filename=*.rst --format="$FLAKE8_FORMAT"
+        RET=$(("${RET}" + $?)) ; echo "${MSG} DONE"
+    else
+        >&2 echo 'Flake8 is not installed.';
+        echo '  To install: $ pip install flake8'
```
Comment (jreback) [inline ci/code_checks.sh]: we don't use pip to install things, rather its conda install

File: ci/code_checks.sh [hunk 5]
```diff
@@ -92,85 +99,94 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
...
     # this particular codebase (e.g. src/headers, src/klib, src/msgpack). However,
     # we can lint all header files since they aren't "generated" like C files are.
-    MSG='Linting .c and .h' ; echo $MSG
-    cpplint --quiet --extensions=c,h --headers=h --recursive --filter=-readability/casting,-runtime/int,-build/include_subdir pandas/_libs/src/*.h pandas/_libs/src/parser pandas/_libs/ujson pandas/_libs/tslibs/src/datetime
-    RET=$(($RET + $?)) ; echo $MSG "DONE"
-
-    echo "isort --version-number"
-    isort --version-number
+    if command -v cpplint >/dev/null; then
+        MSG='Linting .c and .h' ; echo "${MSG}"
```
Comment (jreback) [inline ci/code_checks.sh]: same comment as above

File: ci/code_checks.sh [hunk 6]
```diff
@@ -183,62 +199,65 @@ if mods:
...
-    RET=$(($RET + $?)) ; echo $MSG "DONE"
-
-    MSG='Doctests interval classes' ; echo $MSG
-    pytest --doctest-modules -v \
-        pandas/core/indexes/interval.py \
-        pandas/core/arrays/interval.py \
-        -k"-from_arrays -from_breaks -from_intervals -from_tuples -get_loc -set_closed -to_tuples -interval_range"
-    RET=$(($RET + $?)) ; echo $MSG "DONE"
-
+    if command -v pytest >/dev/null; then
```
Comment (jreback) [inline ci/code_checks.sh]: same comment as above

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by datapythonista — CHANGES_REQUESTED** [review]
-1 on merging this in a single PR

I don't see the value of some of these changes, they seem somehow arbitrary. And we're likely to introduce bugs here.

I think it makes more sense to have each change in a separate PR, so we can discuss them properly.

-1 on checking if `flake8`, `cpplint`... the errors we have now are good enough, and I don't think the extra complexity is justified.

**Comment by datapythonista** [comment]
@mmphego can you split this PR as requested, and fix the conflicts please?

**Comment by mmphego** [comment]
> @mmphego can you split this PR as requested, and fix the conflicts please?

I will be on it, asap - have been on away from my computer. Thanks.

**Pull request closed without merging**